src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/config/ErrorConfiguration.kt

```diff
@@ -1,60 +1,74 @@
-  override fun process(error: HttpError) {
-    when {
-      error.httpStatus.is4xxClientError() -> log.info(prepareMessage(error))
-      error.httpStatus.is5xxServerError() -> log.error(prepareMessage(error))
-    }
-  }
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.ExceptionHandler
+import org.springframework.web.bind.annotation.RestControllerAdvice
+import org.springframework.web.server.ResponseStatusException
+
+data class FieldError(
+  val field: String,
+  val error: String,
+)
 
-  private fun prepareMessage(error: HttpError): String? {
-    return error.errors.map { "${it.message} ${it.arguments.joinToString(" ")}" }.joinToString("; ")
+class ValidationError(override val message: String, val errors: List<FieldError>) : RuntimeException(message)
+
+@RestControllerAdvice
+class ErrorConfiguration {
+  @ExceptionHandler(ValidationError::class)
+  fun handleValidationException(e: ValidationError): ResponseEntity<ErrorResponse> {
+    log.info("validation exception: {}", e.message)
+    return ResponseEntity
+      .status(HttpStatus.BAD_REQUEST)
+      .body(
+        ErrorResponse(
+          status = HttpStatus.BAD_REQUEST.value(),
+          error = "validation error",
+          message = e.message,
+          validationErrors = e.errors,
+        )
+      )
   }
 
-  companion object {
-    private val log = LoggerFactory.getLogger(LoggingErrorWebErrorHandlerPostProcessor::class.java)
+  @ExceptionHandler(ResponseStatusException::class)
+  fun handleResponseException(e: ResponseStatusException): ResponseEntity<ErrorResponse> {
+    log.info("internal exception: {}", e.message)
+    return ResponseEntity
+      .status(e.status)
+      .body(
+        ErrorResponse(
+          status = e.status.value(),
+          error = e.status.reasonPhrase,
+          message = e.reason,
+        )
+      )
   }
-}
 
-@Configuration
-class ErrorsConfiguration {
-  @Bean
-  fun messageSource(): MessageSource? {
-    val messageSource = ReloadableResourceBundleMessageSource()
-    messageSource.setBasename("classpath:errors")
-    messageSource.setDefaultEncoding("UTF-8")
-    return messageSource
+  @ExceptionHandler(java.lang.Exception::class)
+  fun handleException(e: java.lang.Exception): ResponseEntity<ErrorResponse?>? {
+    log.error("unexpected exception: {}", e)
+    return ResponseEntity
+      .status(HttpStatus.INTERNAL_SERVER_ERROR)
+      .body(
+        ErrorResponse(
+          status = HttpStatus.INTERNAL_SERVER_ERROR.value(),
+          error = "unexpected error",
+          message = e.message,
+        )
+      )
   }
 
-  @Bean
-  fun getValidator(): LocalValidatorFactoryBean? {
-    val bean = LocalValidatorFactoryBean()
-    bean.setValidationMessageSource(messageSource())
-    return bean
+  companion object {
+    private val log = LoggerFactory.getLogger(ErrorConfiguration::class.java)
   }
 }
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+data class ErrorResponse(
```

Comment:
this is the key class which describes our error responses

Reply:
Have you considered using the HTTP problem details format for the error responses? The format is almost the same so it might be worth a look. Here is the RFC: https://tools.ietf.org/html/rfc7807

Reply:
thanks for sharing this document. as you mentioned, the format is similar so i think what we have is ok for now.